### PR TITLE
[AMD] Enable FP4 (E2M1) KV cache quantization for NSA with TileLang backend

### PR DESCRIPTION
**Pull request by RolaoDenthu**
## Motivation

Enable FP4 KV cache for NSA on Tilelang backend, based on FP8 attention https://github.com/sgl-project/sglang/pull/21511. This PR should be rebased onto main after #21511 is merged.

with `--mem-fraction-static 0.85`
KV cache data type | max_total_num_tokens
-- | --
BF16 | 2,152,523
FP8 | 3,903,728
MXFP4 | 5,195,188

## Modifications

- Added a Triton quantization kernel, `set_mla_kv_buffer_fp4_quant_kernel`, that fuses BF16→FP4 E2M1 block-wise quantization with paged KV write. Each token is stored as [packed_nope(256B) | block_scales(32B) | rope_bf16(128B)].
- Added a vectorized Triton dequantization kernel, `_dequant_fp4_to_fp8_paged_kernel`, that reads FP4 data and stores FP8 directly with no BF16 intermediate.
- Added `NSATokenToKVPoolFP4` memory pool class managing uint8 storage with FP4 quantization on write.

## Accuracy Tests
GSM8K, 200 questions
| Config | dpskv3.2 | GLM-5 |
|---|---|---|
| FP4 KV cache | 93.5% (187/200) | 95% (190/200) |
| BF16 KV cache | 93.5% (187/200) | 95.5% (191/200) |

GPQA, 198 questions
| Config | dpskv3.2 | GLM-5 |
|---|---|---|
| FP4 KV cache |  76.8% (152/198) | 51.5% (102/198) |
| BF16 KV cache | 75.8% (150/198) | 52.5% (104/198) |

## Speed Tests and Profiling
Start server
```bash
python3 -m sglang.launch_server \
  --model-path /models/DeepSeek-V3.2-Exp/ \
  --trust-remote-code \
  --tensor-parallel-size 8 \
  --nsa-prefill-backend tilelang \
  --nsa-decode-backend tilelang \
  #   --kv-cache-dtype fp4_e2m1 or fp8_e4m3  \
  --mem-fraction-static 0.7 \
  --port 30000
```
Run benchmark
```
python3 -m sglang.bench_serving \
    --host localhost --port 30000 \
    --backend sglang \
    --dataset-name random \
    --random-input 8192 \
    --random-output 1024 \
    --random-range-ratio 1.0 \
    --num-prompts $NP \
    --max-concurrency $CC
```

BF16
Concurrency | Total Throughput (tok/s) | E2E Latency (ms) | TTFT (ms) | ITL (ms)
-- | -- | -- | -- | --
1 | 475 | 19,393 | 1,090 | 17.89
4 | 1,574 | 23,423 | 2,416 | 20.53
16 | 4,086 | 36,083 | 4,288 | 31.08
64 | 7,337 | 80,385 | 16,483 | 62.47

FP8
Concurrency | Total Throughput (tok/s) | E2E Latency (ms) | TTFT (ms) | ITL (ms)
-- | -- | -- | -- | --
1 | 505 (106%) | 18,252 (0.94x) | 497 (0.46x) | 17.36 (0.97x)
4 | 1,706 (108%) | 21,604 (0.92x) | 1,166 (0.48x) | 19.98 (0.97x)
16 | 4,437 (109%) | 33,229 (0.92x) | 3,398 (0.79x) | 29.16 (0.94x)
64 | 8,357 (114%) | 70,570 (0.88x) | 12,541 (0.76x) | 56.72 (0.91x)

FP8 to FP4 (E2M1) KV Cache, comparing to BF16
Concurrency | Total Throughput (tok/s) | E2E Latency (ms) | TTFT (ms) | ITL (ms)
-- | -- | -- | -- | --
1 | 475 (100%) | 19,389 (1.00x) | 659 (0.60x) | 18.31 (1.02x)
4 | 1,597 (101%) | 23,083 (0.99x) | 1,907 (0.79x) | 20.70 (1.01x)
16 | 4,023 (98%) | 36,644 (1.02x) | 5,036 (1.17x) | 30.90 (0.99x)
64 | 7,270 (99%) | 81,110 (1.01x) | 16,789 (1.02x) | 62.88 (1.01x)

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

### CODE REVIEW

**Review by JiwaniZakir — COMMENTED** [review]
In `dequant_fp4_paged_decode`, there's a subtle inconsistency in how invalid (negative) page table entries are handled: the kernel in `_dequant_fp4_to_fp8_paged_kernel` silently clamps negative `src_token_id` values to 0 via `tl.maximum(src_token_id, 0)`, meaning invalid entries will actually read and dequantize real data from slot 0 of the cache into the FP8 output. The `valid_mask` / `torch.where` logic afterward only fixes up `new_page_table_1` but leaves the FP8 values for those entries populated with slot-0 data rather than zeros or a sentinel — this could silently corrupt attention scores if downstream masking isn't perfectly aligned.

Additionally, in `dequantize_fp4_to_fp8_paged`, when a pre-allocated `output` buffer is provided but `output.shape[0] < num_tokens`, the function silently falls back to allocating a fresh tensor. The docstring on `dequant_fp4_paged_decode` explicitly calls this "CUDA-graph safe" and emphasizes pre-allocated buffers to avoid per-step allocations, but this silent fallback would break CUDA graph capture without any warning. An explicit assertion or at minimum a logged warning would be safer here.

**Review by HaiShaw — COMMENTED** [review]
Conflict to solve. Possible to add GPQA accuracy scores too?

File: python/sglang/srt/models/deepseek_common/attention_forward_methods/forward_mla.py [hunk 1]
```diff
@@ -398,7 +398,7 @@ def forward_absorb_core(
                     ),
                 )
         else:
-            if _use_aiter_gfx95:
+            if _use_aiter_gfx95 and self.kv_cache_dtype != "fp4_e2m1":
```
Comment (HaiShaw) [inline python/sglang/srt/models/deepseek_common/attention_forward_methods/forward_mla.py:408]: why `and self.kv_cache_dtype != "fp4_e2m1"`

File: python/sglang/srt/models/deepseek_common/attention_forward_methods/forward_mla.py [hunk 2]
```diff
@@ -592,4 +592,5 @@ def _skip_rope_for_nsa_tilelang_fused(self: DeepseekV2AttentionMLA) -> bool:
                 server_args.nsa_decode_backend == "tilelang"
                 or server_args.nsa_prefill_backend == "tilelang"
             )
+            and self.kv_cache_dtype != "fp4_e2m1"
```
Comment (HaiShaw) [inline python/sglang/srt/models/deepseek_common/attention_forward_methods/forward_mla.py:609]: why `and self.kv_cache_dtype != "fp4_e2m1"`

**Review by HaiShaw — COMMENTED** [review]
https://github.com/sgl-project/sglang/pull/21511 merged, rebase?

File: python/sglang/srt/mem_cache/utils.py [hunk 3]
```diff
@@ -322,6 +322,163 @@ def get_mla_kv_buffer_triton(
     )
 
 
+FP4_BLOCK_SIZE = 16
```
Comment (HaiShaw) [inline python/sglang/srt/mem_cache/utils.py]: For NVFP4 is 16, for MxFP4 is 32

**Review by HaiShaw — CHANGES_REQUESTED** [review]
The implementation mixed NVFP4 and MxFP4 in following aspects:
- block size, uses 16 (NVFP4)
- scaling, uses MxFP4 definition

Please choose one, just for KV cache, either should work (till later we use FP4 matmul) but please be consistent.

File: python/sglang/srt/models/deepseek_common/attention_forward_methods/forward_mla.py (hunk 2 above)
Comment (RolaoDenthu) [reply python/sglang/srt/models/deepseek_common/attention_forward_methods/forward_mla.py:609]: These two will lead to aiter's fused_qk_rope_cat_and_cache_mla kernel, which fuses KV cache writing inside, but it does not support writing FP4 KV cache. For FP4, we skip this and go through set_mla_kv_buffer instead, which calls our own Triton quantization kernel to write the FP4 layout.

**Comment by RolaoDenthu** [comment]
> The implementation mixed NVFP4 and MxFP4 in following aspects:
> 
> * block size, uses 16 (NVFP4)
> * scaling, uses MxFP4 definition
> 
> Please choose one, just for KV cache, either should work (till later we use FP4 matmul) but please be consistent.

Thank you for review! I have changed the block size and updated the benchmark statistics.
